### ISSUE TITLE
Await in tracing span

### ISSUE THREAD

**Issue by llgerard** [rust-lang/rust-clippy#8722]
### What it does

Check that while entered in a span, calls to `.await` are properly instrumented.

Extract from the documentation https://docs.rs/tracing/latest/tracing/struct.Span.html#in-asynchronous-code :

> Warning: in asynchronous code that uses [async/await syntax](https://rust-lang.github.io/async-book/01_getting_started/04_async_await_primer.html), Span::enter should be used very carefully or avoided entirely. Holding the drop guard returned by Span::enter across .await points will result in incorrect traces.

### Lint Name

uninstrumented_await_while_span_entered

### Category

correctness, suspicious

### Advantage

Correct span usage. Without instrumentation of the async task, it will wrongly report the current span.

### Example

```rust
async fn my_async_function() {
    let span = info_span!("my_async_function");
    let _enter = span.enter();
    // The `await` keyword may yield, causing the runtime
    // to switch to another task, while remaining in this span!
    some_other_async_function().await

    // ...
}
```

Could be written as:

```rust
use tracing::Instrument;

async fn my_async_function() {
    let span = info_span!("my_async_function");
    async move {
       some_other_async_function().await;
       //...
    }
      // instrument the async block with the span
      .instrument(span)
      .await
}
```

or less generally as
```rust
use tracing::Instrument;

async fn my_async_function() {
    let some_value = some_other_async_function()
       .instrument(debug_span!("my_async_function"))
       .await;

    // ...
}
```

**Comment by llgerard** [comment]
One could do a rough check with `await-holding-invalid-type` but that wouldn't allow correctly instrumented async tasks.

**Comment by arielb1** [comment]
> One could do a rough check with await-holding-invalid-type but that wouldn't allow correctly instrumented async tasks.

What's the problem with using `await-holding-invalid-type`?
